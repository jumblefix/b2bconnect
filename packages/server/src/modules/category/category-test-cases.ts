const name = 'Information Technology';
const child = 'Child';

export const addCategoryWoLogin = {
  caseId: 'add category without login',
  query: `mutation { addCategory(name:"${name}"){
          id
          name
          slug
        }}`,
  session: {
    userId: '',
    isAdmin: false,
  },
};

export const addCategoryWoAdmin = {
  caseId: 'logged in but not admin',
  query: `mutation { addCategory(name:"${name}"){
          id
          name
          slug
        }}`,
  session: {
    userId: '123456',
    isAdmin: false,
  },
};

export const addCategoryValid = {
  caseId: 'addCategoryValid',
  query: `mutation { addCategory(name:"${name}"){
          id
          name
          slug
        }}`,
  session: {
    userId: '123456',
    isAdmin: true,
  },
};

export const addCategoryWParent = {
  caseId: 'add category with parent',
  query: `mutation { addCategory(name:"${child}", parentId:1 ){
          id
          name
          slug
        }}`,
  session: {
    userId: '123456',
    isAdmin: true,
  },
};

export const addCategoryInvalidParent = {
  caseId: 'add category with invalid parent',
  query: `mutation { addCategory(name:"${child}", parentId:0 ){
          id
          name
          slug
        }}`,
  session: {
    userId: '123456',
    isAdmin: true,
  },
};

export const listMainCategories = {
  caseId: 'listMainCategories',
  query: `{ listMainCategories{ slug } }`,
  session: {},
};

export const getCategoryById = {
  caseId: 'getCategoryById',
  query: `{ getCategoryById(id:"1"){ slug } }`,
  session: {},
};

export const getBreadCrumbPath = {
  caseId: 'getBreadCrumbPath',
  query: `{ getBreadCrumbPath(id:"2"){ slug, parent{
        slug
      } } }`,
  session: {},
};

export const getBreadCrumbPath2 = {
  caseId: 'getBreadCrumbPath2',
  query: `{ getBreadCrumbPath(id:"0"){ slug, parent{
        slug
      } } }`,
  session: {},
};

export const getChildCategories = {
  caseId: 'getChildCategories',
  query: `{ getChildCategories(id:"1"){ slug children{ slug } } }`,
  session: {},
};

export const getChildCategories2 = {
  caseId: 'getChildCategories2',
  query: `{ getChildCategories(id:"0"){ slug children{ slug } } }`,
  session: {},
};
