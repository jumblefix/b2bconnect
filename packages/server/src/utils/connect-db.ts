// tslint:disable-next-line:no-var-requires
const iconv = require('iconv-lite');
// tslint:disable-next-line:no-var-requires
const encodings = require('iconv-lite/encodings');
iconv.encodings = encodings;

import { createConnection, getConnectionOptions } from 'typeorm';

export const connectDb = async () => {
  const connectionOptions = await getConnectionOptions(process.env.NODE_ENV);

  return createConnection({
    ...connectionOptions,
    name: 'default',
  });
};

export const connectDbTest = async (resetDB: boolean = false) => {
  const connectionOptions = await getConnectionOptions(process.env.NODE_ENV);

  return createConnection({
    ...connectionOptions,
    name: 'default',
    synchronize: resetDB,
    dropSchema: resetDB,
  });
};
