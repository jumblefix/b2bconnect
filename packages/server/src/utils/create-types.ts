import { generateNamespace } from '@gql2ts/from-schema';
import * as fs from 'fs';
import * as path from 'path';
import { genSchema } from './schema-utils';

const typescriptTypes = generateNamespace('GQL', genSchema());

fs.writeFile(
  path.join(__dirname, '../typings/GQL/index.d.ts'),
  typescriptTypes,
  err => {
    console.error(err);
  },
);
