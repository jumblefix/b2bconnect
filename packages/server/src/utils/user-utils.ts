import { Request, Response } from 'express';
import * as IORedis from 'ioredis';
import { v4 } from 'uuid';
import { ValidationError } from 'yup';
import {
  FORGOT_PASSWORD_PREFIX,
  REDIS_SESSION_PREFIX,
  TokenTypes,
  USER_SESSION_PREFIX,
} from '../constants';
import { User } from '../entity/User';
import { GraphQLMiddlewareFunc, Resolver } from '../typings/app-utility-types';

export const createForgotPasswordLink = async (
  url: string,
  userId: string,
  redis: IORedis.Redis,
) => {
  const id = v4();
  await redis.set(`${FORGOT_PASSWORD_PREFIX}${id}`, userId, 'ex', 60 * 60);
  return `${url}/change-password/${id}`;
};

export const removeAllUsersSessions = async (
  userId: string,
  redis: IORedis.Redis,
) => {
  const sessionIds = await redis.lrange(
    `${USER_SESSION_PREFIX}${userId}`,
    0,
    -1,
  );

  const promises = [];
  // tslint:disable-next-line:prefer-for-of
  for (let i = 0; i < sessionIds.length; i += 1) {
    promises.push(redis.del(`${REDIS_SESSION_PREFIX}${sessionIds[i]}`));
  }
  await Promise.all(promises);
};

export const formatYupError = (err: ValidationError) => {
  const errors: [{path: string; message: string}] = [] as any;
  err.inner.forEach(e => {
    errors.push({
      path: e.path,
      message: e.message,
    });
  });

  return errors;
};

export const createMiddleware = (
  middlewareFunc: GraphQLMiddlewareFunc,
  resolverFunc: Resolver,
) => (parent: any, args: any, context: any, info: any) =>
  middlewareFunc(resolverFunc, parent, args, context, info);

export const middleware = async (
  resolver: Resolver,
  parent: any,
  args: any,
  context: any,
  info: any,
) => {
  return resolver(parent, args, context, info);
};

export const createTokenLink = async (
  url: string,
  userId: string,
  redis: IORedis.Redis,
  type: TokenTypes,
) => {
  const id = v4();
  await redis.set(id, userId, 'ex', 60 * 60 * 24);
  return `${url}/${type}/${id}`;
};

export const confirmEmail = async (req: Request, res: Response) => {
  const { id } = req.params;
  const RedisInstance = new IORedis();
  const userId = await RedisInstance.get(id);
  if (userId) {
    await User.update({ id: userId }, { confirmed: true });
    await RedisInstance.del(id);
    res.redirect(`${process.env.FRONTEND_HOST}/login`);
  } else {
    res.json({ message: 'invalid or expired register token' });
  }
};

export const resetPassword = async (req: Request, res: Response) => {
  const { id } = req.params;
  const RedisInstance = new IORedis();
  const userId = await RedisInstance.get(id);
  if (userId) {
    res.redirect(`${process.env.FRONTEND_HOST}/reset/${id}`);
  } else {
    res.json({ message: 'invalid or expired reset token' });
  }
};
