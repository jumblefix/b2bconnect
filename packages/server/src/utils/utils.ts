import {
  formatError as formatApolloError,
  isInstance as isApolloErrorInstance,
} from 'apollo-errors';
import * as nodemailer from 'nodemailer';
import slugify from 'slugify';
import { ValidationError } from 'yup';
import { Env } from '../constants';
import {
  AuthenticationError,
  AuthorizationError,
  InputValidationError,
} from './errors';

export const add = (...args: number[]) => args.reduce((c, p) => c + p);

export const formatYupError = (err: ValidationError) => {
  const errors: [{ path: string; message: string }] = [] as any;
  err.inner.forEach(e => {
    errors.push({
      path: e.path,
      message: e.message,
    });
  });
  return errors;
};

export function formatError(error: any) {
  const { originalError } = error;
  if (isApolloErrorInstance(originalError)) {
    // log internalData to stdout but not include it in the formattedError
    console.error(
      JSON.stringify({
        type: `error`,
        data: originalError.data,
        internalData: originalError.internalData,
      }),
    );
  }
  return formatApolloError(error);
}

export const makeSlug = (str: string) => slugify(str, { lower: true });

export const sendEmail = async ({
  to,
  subject,
  text,
  html,
}: SendEmailParams) => {
  let mailConfig: any;
  if (process.env.NODE_ENV === Env.production) {
    mailConfig = {
      host: 'smtp.sendgrid.net',
      port: 587,
      auth: {
        user: 'real.user',
        pass: 'verysecret',
      },
    };
  } else {
    mailConfig = {
      host: process.env.MAIL_HOST,
      port: process.env.MAIL_PORT,
      auth: {
        user: process.env.MAIL_USER,
        pass: process.env.MAIL_PASS,
      },
    };
  }

  const transport = nodemailer.createTransport(mailConfig);

  try {
    const info = await transport.sendMail({
      to,
      subject,
      text,
      html,
    });

    console.log(nodemailer.getTestMessageUrl(info));
  } catch (error) {
    console.error(error);
    throw new Error(error);
  }
  return true;
};

export const checkAdminRights = (session: any) => {
  const { userId, isAdmin } = session;
  if (!userId) {
    throw new AuthenticationError();
  }
  if (!isAdmin) {
    throw new AuthorizationError();
  }
};

export const validateInputs = async (schema: any, inputs: any) => {
  try {
    await schema.validate(inputs, { abortEarly: false });
  } catch (err) {
    const errors = formatYupError(err);
    throw new InputValidationError({
      data: errors,
    });
  }
};
