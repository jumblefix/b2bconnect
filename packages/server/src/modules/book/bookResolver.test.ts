import { graphql } from 'graphql';
import { Connection } from 'typeorm';
import { genSchema } from '../../utils/schema-utils';
import {
  addBookInvalidCategory,
  addBookInvalidData,
  addBookValidData,
  getBook,
  getBookByCategory,
  getBookByCategoryInvalid,
  listBooks,
} from './book-test-cases';

import { addCategoryValid } from '../category/category-test-cases';
import { connectDbTest } from './../../utils/connect-db';

let connection: Connection;

beforeAll(async () => {
  connection = await connectDbTest(true);
});

afterAll(async () => {
  connection.close();
});

describe('books resolver', () => {
  const cases = [
    addBookInvalidCategory,
    addBookInvalidData,
    addCategoryValid,
    addBookValidData,
    listBooks,
    getBookByCategoryInvalid,
    getBookByCategory,
    getBook,
  ];
  cases.forEach(c => {
    const { query, session } = c;
    it(`case: ${c.caseId}`, async () => {
      const ctx = { session, db: connection };
      const result = await graphql(genSchema(), query, null, ctx, {});
      expect(result).toMatchSnapshot();
      if (result.errors && result.errors!.length) {
        // console.log(result.errors);
      }
    });
  });
});
