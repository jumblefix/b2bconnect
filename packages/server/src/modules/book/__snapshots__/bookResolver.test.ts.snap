// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`books resolver case: add book with invalid category 1`] = `
Object {
  "data": Object {
    "addBook": null,
  },
  "errors": Array [
    [GraphQLError: Invalid Category],
  ],
}
`;

exports[`books resolver case: add book with invalid data 1`] = `
Object {
  "data": Object {
    "addBook": null,
  },
  "errors": Array [
    [GraphQLError: Validation failed],
  ],
}
`;

exports[`books resolver case: addBookValidData 1`] = `
Object {
  "data": Object {
    "addBook": Object {
      "category": Object {
        "name": "Information Technology",
      },
      "isbn": "1233434234",
      "title": "My Book",
    },
  },
}
`;

exports[`books resolver case: addCategoryValid 1`] = `
Object {
  "data": Object {
    "addCategory": Object {
      "id": "1",
      "name": "Information Technology",
      "slug": "information-technology",
    },
  },
}
`;

exports[`books resolver case: getBook 1`] = `
Object {
  "data": Object {
    "getBook": null,
  },
}
`;

exports[`books resolver case: getBookByCategory 1`] = `
Object {
  "data": Object {
    "getBookByCategory": null,
  },
  "errors": Array [
    [GraphQLError: Invalid Category],
  ],
}
`;

exports[`books resolver case: getBookByCategory 2`] = `
Object {
  "data": Object {
    "getBookByCategory": Array [
      Object {
        "title": "My Book",
      },
    ],
  },
}
`;

exports[`books resolver case: listBooks 1`] = `
Object {
  "data": Object {
    "listBooks": Array [
      Object {
        "isbn": "1233434234",
        "title": "My Book",
      },
    ],
  },
}
`;
